btree/btree.c

```diff
@@ -3460,46 +3379,95 @@ static int64_t btree_put_kv_tick(struct m0_sm_op *smop)
 				node_unlock(lev->l_node);
 				if (oi->i_key_found)
 					return P_LOCK;
-				return P_ALLOC;
+				/**
+				 * Initialize i_alloc_lev to level of leaf
+				 * node.
+				 */
+				oi->i_alloc_lev = oi->i_used;
+				return P_ALLOC_REQUIRE;
 			}
 		} else {
 			node_op_fini(&oi->i_nop);
 			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
 		}
-	case P_ALLOC: {
-		bool alloc = false;
+	case P_ALLOC_REQUIRE:{
 		do {
-			lev = &oi->i_level[oi->i_used];
-			/**
-			 * Validate node to dertmine if lev->l_node is still
-			 * exists.
-			 */
-			if (!node_isvalid(lev->l_node)) {
-				oi->i_used = oi->i_height - 1;
+			lev = &oi->i_level[oi->i_alloc_lev];
+			if (!node_isvalid(lev->l_node))
 				return m0_sm_op_sub(&bop->bo_op, P_CLEANUP,
 						    P_SETUP);
-			}
+
 			if (!node_isoverflow(lev->l_node))
 				break;
-			if (oi->i_used == 0) {
-				if (lev->l_alloc == NULL ||
-				    oi->i_extra_node == NULL)
-					alloc = true;
-				break;
-			} else if (lev->l_alloc == NULL) {
-				alloc = true;
-				break;
-			}
-
-			oi->i_used--;
-		} while (1);
+			if (lev->l_alloc == NULL || (oi->i_alloc_lev == 0 &&
+			    oi->i_extra_node == NULL)) {
+				/**
+				 * Depending on the level of node, shift can be
+				 * updated.
+				 */
+				int ksize = node_keysize(lev->l_node);
+				int vsize = node_valsize(lev->l_node);
+				int shift = node_shift(lev->l_node);
+				oi->i_nop.no_opc = NOP_ALLOC;
+				return node_alloc(&oi->i_nop, tree,
+						  shift, lev->l_node->n_type,
+						  ksize, vsize, bop->bo_tx,
+						  P_ALLOC_STORE);
 
-		if (alloc)
-			return btree_put_alloc_phase(bop);
-		/* Reset oi->i_used */
-		oi->i_used = oi->i_height - 1;
+			}
+			oi->i_alloc_lev--;
+		} while (oi->i_alloc_lev >= 0);
 		return P_LOCK;
 	}
+	case P_ALLOC_STORE: {
+		lev = &oi->i_level[oi->i_alloc_lev];
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			if (oi->i_alloc_lev == 0) {
```

Comment:
You can rearrange the code as below to avoid multiple level of indentation.

```
if (oi->i_nop.no_op.o_sm.sm_rc != 0) {
                      if (lock_acquired)
				lock_op_unlock(tree);
			node_op_fini(&oi->i_nop);
			return fail(bop, oi->i_nop.no_op.o_sm.sm_rc);
}
lev = &oi->i_level[oi->i_alloc_lev];
if (oi->i_alloc_lev == 0) {
     ------
     ------
}

Reply:
done